File: examples/java/cases/tasks/task_move/TaskMoveDomainCase.java
package cases.tasks.task_move;

import core.BaseDomainCase;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class TaskMoveDomainCase
    extends BaseDomainCase<TaskMoveDomainCase.TaskMoveInput, TaskMoveDomainCase.TaskMoveOutput> {
  public static final List<String> TASK_STATUS_VALUES = List.of("todo", "doing", "done");

  public record Task(
      String id,
      String title,
      String description,
      String status,
      String createdAt,
      String updatedAt
  ) {
    public static Task fromMap(Map<String, Object> value, String source) {
      if (value == null) {
        throw new IllegalArgumentException(source + " must be an object");
      }

      String id = requiredString(value, "id", source);
      String title = requiredString(value, "title", source);
      String status = requiredString(value, "status", source);
      String createdAt = requiredString(value, "createdAt", source);
      String updatedAt = requiredString(value, "updatedAt", source);
      Object description = value.get("description");

      if (!TASK_STATUS_VALUES.contains(status)) {
        throw new IllegalArgumentException(
            source + ".status must be one of " + String.join(", ", TASK_STATUS_VALUES)
        );
      }

      if (description != null && !(description instanceof String)) {
        throw new IllegalArgumentException(source + ".description must be a string when provided");
      }

      return new Task(id, title, (String) description, status, createdAt, updatedAt);
    }

    public Map<String, Object> toMap() {
      Map<String, Object> value = new LinkedHashMap<>();
      value.put("id", id);
      value.put("title", title);
      if (description != null && !description.isBlank()) {
        value.put("description", description);
      }
      value.put("status", status);
      value.put("createdAt", createdAt);
      value.put("updatedAt", updatedAt);
      return value;
    }
  }

  public record TaskMoveInput(String taskId, String targetStatus) {
    public static TaskMoveInput fromMap(Map<String, Object> value) {
      if (value == null) {
        throw new IllegalArgumentException("input must be an object");
      }

      Object taskId = value.get("taskId");
      Object targetStatus = value.get("targetStatus");

      if (!(taskId instanceof String taskIdValue)) {
        throw new IllegalArgumentException("taskId is required");
      }

      if (!(targetStatus instanceof String targetStatusValue)) {
        throw new IllegalArgumentException("targetStatus is required");
      }

      return new TaskMoveInput(taskIdValue, targetStatusValue);
    }
  }

  public record TaskMoveOutput(Task task) {
    public static TaskMoveOutput fromMap(Map<String, Object> value) {
      if (value == null || !(value.get("task") instanceof Map<?, ?> taskMap)) {
        throw new IllegalArgumentException("task_move output must contain task");
      }

      @SuppressWarnings("unchecked")
      Map<String, Object> taskValue = (Map<String, Object>) taskMap;
      return new TaskMoveOutput(Task.fromMap(taskValue, "task_move.output.task"));
    }

    public Map<String, Object> toMap() {
      return Map.of("task", task.toMap());
    }
  }

  @Override
  public String caseName() {
    return "task_move";
  }

  @Override
  public String description() {
    return "Moves an existing task card to another board column.";
  }

  @Override
  public AppSchema inputSchema() {
    return AppSchema.object()
        .property("taskId", AppSchema.string().description("Identifier of the task that will be moved."))
        .property("targetStatus", AppSchema.string()
            .description("Destination board column for the task.")
            .enumValues("todo", "doing", "done"))
        .required("taskId", "targetStatus")
        .additionalProperties(false);
  }

  @Override
  public AppSchema outputSchema() {
    return AppSchema.object()
        .property("task", AppSchema.object()
            .property("id", AppSchema.string())
            .property("title", AppSchema.string())
            .property("description", AppSchema.string())
            .property("status", AppSchema.string().enumValues("todo", "doing", "done"))
            .property("createdAt", AppSchema.string())
            .property("updatedAt", AppSchema.string())
            .required("id", "title", "status", "createdAt", "updatedAt")
            .additionalProperties(false))
        .required("task")
        .additionalProperties(false);
  }

  @Override
  public void validate(TaskMoveInput input) {
    if (input == null) {
      throw new IllegalArgumentException("input must be an object");
    }

    if (input.taskId() == null || input.taskId().trim().isEmpty()) {
      throw new IllegalArgumentException("taskId is required");
    }

    if (!TASK_STATUS_VALUES.contains(input.targetStatus())) {
      throw new IllegalArgumentException(
          "targetStatus must be one of " + String.join(", ", TASK_STATUS_VALUES)
      );
    }
  }

  public void validateOutput(TaskMoveOutput output) {
    if (output == null) {
      throw new IllegalArgumentException("output must be an object");
    }

    if (!TASK_STATUS_VALUES.contains(output.task().status())) {
      throw new IllegalArgumentException("task_move.output.task.status is invalid");
    }
  }

  @Override
  public List<String> invariants() {
    return List.of(
        "Moving a task never changes its identity.",
        "A move only updates status and, when applicable, updatedAt.",
        "Moving to the same status is idempotent and returns the unchanged task."
    );
  }

  @Override
  public List<DomainExample<TaskMoveInput, TaskMoveOutput>> examples() {
    DomainExample<TaskMoveInput, TaskMoveOutput> moveTodoToDoing = new DomainExample<>();
    moveTodoToDoing.name = "move_todo_to_doing";
    moveTodoToDoing.description = "A task leaves todo and enters doing.";
    moveTodoToDoing.input = new TaskMoveInput("task_001", "doing");
    moveTodoToDoing.output = new TaskMoveOutput(
        new Task(
            "task_001",
            "Ship the Java example",
            null,
            "doing",
            "2026-03-18T12:00:00.000Z",
            "2026-03-18T12:20:00.000Z"
        )
    );

    DomainExample<TaskMoveInput, TaskMoveOutput> idempotentMove = new DomainExample<>();
    idempotentMove.name = "idempotent_move";
    idempotentMove.description = "Moving to the same status keeps the task unchanged.";
    idempotentMove.input = new TaskMoveInput("task_002", "done");
    idempotentMove.output = new TaskMoveOutput(
        new Task(
            "task_002",
            "Prepare release notes",
            null,
            "done",
            "2026-03-18T12:10:00.000Z",
            "2026-03-18T12:10:00.000Z"
        )
    );

    return List.of(moveTodoToDoing, idempotentMove);
  }

  @Override
  public void test() throws Exception {
    validate(new TaskMoveInput("task_001", "doing"));

    boolean rejected = false;
    try {
      validate(new TaskMoveInput("", "doing"));
    } catch (IllegalArgumentException ignored) {
      rejected = true;
    }

    if (!rejected) {
      throw new IllegalStateException("test: validate must reject empty taskId");
    }

    rejected = false;
    try {
      validate(new TaskMoveInput("task_001", "invalid"));
    } catch (IllegalArgumentException ignored) {
      rejected = true;
    }

    if (!rejected) {
      throw new IllegalStateException("test: validate must reject invalid targetStatus");
    }

    validateOutput(new TaskMoveOutput(
        new Task(
            "task_001",
            "Valid task",
            null,
            "doing",
            "2026-03-18T12:00:00.000Z",
            "2026-03-18T12:20:00.000Z"
        )
    ));
  }

  private static String requiredString(Map<String, Object> value, String field, String source) {
    Object current = value.get(field);
    if (!(current instanceof String stringValue) || stringValue.trim().isEmpty()) {
      throw new IllegalArgumentException(source + "." + field + " must be a non-empty string");
    }

    return stringValue;
  }
}

File: examples/go/cases/tasks/task_move/task_move.domain.case.go
package task_move

import (
	"encoding/json"
	"fmt"
	"strings"

	"app-protocol/examples/go/core"
)

type TaskStatus string

const (
	TaskStatusTodo  TaskStatus = "todo"
	TaskStatusDoing TaskStatus = "doing"
	TaskStatusDone  TaskStatus = "done"
)

type Task struct {
	ID          string     `json:"id"`
	Title       string     `json:"title"`
	Description string     `json:"description,omitempty"`
	Status      TaskStatus `json:"status"`
	CreatedAt   string     `json:"createdAt"`
	UpdatedAt   string     `json:"updatedAt"`
}

type TaskMoveInput struct {
	TaskID       string     `json:"taskId"`
	TargetStatus TaskStatus `json:"targetStatus"`
}

type TaskMoveOutput struct {
	Task Task `json:"task"`
}

type Domain struct {
	core.DomainCaseSupport
}

func NewDomain() *Domain {
	return &Domain{}
}

func (d *Domain) CaseName() string {
	return "task_move"
}

func (d *Domain) Description() string {
	return "Moves an existing task card to another board column."
}

func (d *Domain) InputSchema() core.AppSchema {
	return core.AppSchema{
		"type": "object",
		"properties": map[string]any{
			"taskId": map[string]any{
				"type":        "string",
				"description": "Identifier of the task that will be moved.",
			},
			"targetStatus": map[string]any{
				"type":        "string",
				"description": "Destination board column for the task.",
				"enum":        []string{"todo", "doing", "done"},
			},
		},
		"required":             []string{"taskId", "targetStatus"},
		"additionalProperties": false,
	}
}

func (d *Domain) OutputSchema() core.AppSchema {
	return core.AppSchema{
		"type": "object",
		"properties": map[string]any{
			"task": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":          map[string]any{"type": "string"},
					"title":       map[string]any{"type": "string"},
					"description": map[string]any{"type": "string"},
					"status": map[string]any{
						"type": "string",
						"enum": []string{"todo", "doing", "done"},
					},
					"createdAt": map[string]any{"type": "string"},
					"updatedAt": map[string]any{"type": "string"},
				},
				"required":             []string{"id", "title", "status", "createdAt", "updatedAt"},
				"additionalProperties": false,
			},
		},
		"required":             []string{"task"},
		"additionalProperties": false,
	}
}

func (d *Domain) ParseInput(value any) (TaskMoveInput, error) {
	record, ok := value.(map[string]any)
	if !ok {
		encoded, err := json.Marshal(value)
		if err != nil {
			return TaskMoveInput{}, fmt.Errorf("input must be an object")
		}
		if err := json.Unmarshal(encoded, &record); err != nil {
			return TaskMoveInput{}, fmt.Errorf("input must be an object")
		}
	}

	input := TaskMoveInput{}
	if taskID, ok := record["taskId"].(string); ok {
		input.TaskID = taskID
	}
	if targetStatus, ok := record["targetStatus"].(string); ok {
		input.TargetStatus = TaskStatus(targetStatus)
	}

	return input, d.Validate(input)
}

func (d *Domain) Validate(input TaskMoveInput) error {
	if strings.TrimSpace(input.TaskID) == "" {
		return fmt.Errorf("taskId is required")
	}

	switch input.TargetStatus {
	case TaskStatusTodo, TaskStatusDoing, TaskStatusDone:
	default:
		return fmt.Errorf("targetStatus must be one of todo, doing, done")
	}

	return nil
}

func (d *Domain) ValidateTask(task Task) error {
	if strings.TrimSpace(task.ID) == "" {
		return fmt.Errorf("id must be a non-empty string")
	}

	if strings.TrimSpace(task.Title) == "" {
		return fmt.Errorf("title must be a non-empty string")
	}

	switch task.Status {
	case TaskStatusTodo, TaskStatusDoing, TaskStatusDone:
	default:
		return fmt.Errorf("status must be one of todo, doing, done")
	}

	if strings.TrimSpace(task.CreatedAt) == "" {
		return fmt.Errorf("createdAt must be a non-empty string")
	}

	if strings.TrimSpace(task.UpdatedAt) == "" {
		return fmt.Errorf("updatedAt must be a non-empty string")
	}

	return nil
}

func (d *Domain) ValidateOutput(output TaskMoveOutput) error {
	return d.ValidateTask(output.Task)
}

func (d *Domain) Invariants() []string {
	return []string{
		"Moving a task never changes its identity.",
		"A move only updates status and, when applicable, updatedAt.",
		"Moving to the same status is idempotent and returns the unchanged task.",
	}
}

func (d *Domain) Examples() []core.DomainExample {
	return []core.DomainExample{
		{
			Name:        "move_todo_to_doing",
			Description: "A task leaves todo and enters doing.",
			Input: TaskMoveInput{
				TaskID:       "task_001",
				TargetStatus: TaskStatusDoing,
			},
			Output: TaskMoveOutput{
				Task: Task{
					ID:        "task_001",
					Title:     "Ship the Go example",
					Status:    TaskStatusDoing,
					CreatedAt: "2026-03-18T12:00:00.000Z",
					UpdatedAt: "2026-03-18T12:20:00.000Z",
				},
			},
		},
		{
			Name:        "idempotent_move",
			Description: "Moving to the same status keeps the task unchanged.",
			Input: TaskMoveInput{
				TaskID:       "task_002",
				TargetStatus: TaskStatusDone,
			},
			Output: TaskMoveOutput{
				Task: Task{
					ID:        "task_002",
					Title:     "Prepare release notes",
					Status:    TaskStatusDone,
					CreatedAt: "2026-03-18T12:10:00.000Z",
					UpdatedAt: "2026-03-18T12:10:00.000Z",
				},
			},
		},
	}
}

func (d *Domain) Definition() core.DomainDefinition {
	return core.DomainDefinition{
		CaseName:     d.CaseName(),
		Description:  d.Description(),
		InputSchema:  d.InputSchema(),
		OutputSchema: d.OutputSchema(),
		Invariants:   d.Invariants(),
		ValueObjects: d.ValueObjects(),
		Enums:        d.Enums(),
		Examples:     d.Examples(),
	}
}

func (d *Domain) Test() error {
	if err := d.Validate(TaskMoveInput{
		TaskID:       "task_001",
		TargetStatus: TaskStatusDoing,
	}); err != nil {
		return err
	}

	if err := d.Validate(TaskMoveInput{
		TaskID:       "",
		TargetStatus: TaskStatusDoing,
	}); err == nil {
		return fmt.Errorf("test: validate must reject empty taskId")
	}

	if err := d.Validate(TaskMoveInput{
		TaskID:       "task_001",
		TargetStatus: "invalid",
	}); err == nil {
		return fmt.Errorf("test: validate must reject invalid targetStatus")
	}

	if err := d.ValidateOutput(TaskMoveOutput{
		Task: Task{
			ID:        "task_001",
			Title:     "Valid task",
			Status:    TaskStatusDoing,
			CreatedAt: "2026-03-18T12:00:00.000Z",
			UpdatedAt: "2026-03-18T12:20:00.000Z",
		},
	}); err != nil {
		return err
	}

	return nil
}
